### PR TITLE
AzureUSGovernment and Rec engine changes

### PR DESCRIPTION
**Pull request by Akreddy123**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK/Framework/Models/AzSKSettings.ps1 [hunk 1]
```diff
@@ -28,7 +28,8 @@ class AzSKSettings {
 	[TertiaryBool] $AllowSelfSignedWebhookCertificate;
 	[bool] $EnableAADAuthForOnlinePolicyStore;
     [bool] $UseOnlinePolicyStore;
-    [string] $OnlinePolicyStoreUrl;
+	[string] $OnlinePolicyStoreUrl;
+	[string] $AzureEnvironment;
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Models/AzSKSettings.ps1]: Should we have default value to "AzureCloud"?

File: src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 2]
```diff
@@ -2987,6 +2987,12 @@ class CCAutomation: CommandBase
 		$CoreSetupSrcUrl = [ConfigurationManager]::GetAzSKConfigData().CASetupRunbookURL.Replace('$',"``$")
 		$AzSKCARunbookVersion = [ConfigurationManager]::GetAzSKConfigData().AzSKCARunbookVersion
 		$telemetryKey = ""
+		$AzureEnv = [ConfigurationManager]::GetAzSKSettings().AzureEnvironment
+		if([string]::IsNullOrWhiteSpace($AzureEnv))
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1]: We should also check if we get default string "[#AzureEnvironment#]" in $AzureEnv value. And set to AzureCloud...

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 [hunk 3]
```diff
@@ -139,12 +141,24 @@ class AppService: SVTBase
 					}
 					else
 					{
+						if([Helpers]::AzureEnv.AzureEnvironment = "AzureUSGovernment")
```
[the remark this answers is not in the document]
Comment (sbyna-msft) [reply src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: use control settings instead

File: src/AzSK/Framework/Abstracts/AzSKRoot.ps1 [hunk 4]
```diff
@@ -28,7 +28,8 @@ class AzSKRoot: EventBase
     
     hidden [void] SetAzureContext()
     {
-		$currentContext = [Helpers]::GetCurrentRMContext()
+		$AzureEnv = [ConfigurationManager]::GetLocalAzSKSettings()
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Abstracts/AzSKRoot.ps1]: Does this PR covers Contoso(Custom Policy) related changes as well? As it has multiple refences related to AzureEnv

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 [hunk 5]
```diff
@@ -465,11 +479,41 @@ class AppService: SVTBase
...
+				if([Helpers]::CheckMember($this.WebAppDetails,"EnabledHostNames"))
+				{
+					if((($this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('scm') }) | Measure-Object).Count -eq 1)
+					{
+						$scmURL = $this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('scm') }
+						$apiFunctionsUrl = [string]::Format("https://{0}/api/functions",$scmURL)
+					}
+					else
+					{
+						if([Helpers]::AzureEnv.AzureEnvironment = "AzureUSGovernment")
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: AzureUSGovernment is hardcoded here. Also seems like if and else has same statements

File: src/AzSK/Framework/Models/AzSKSettings.ps1 (hunk 1 above)
Comment (Akreddy123) [reply src/AzSK/Framework/Models/AzSKSettings.ps1]: Yes

File: src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1 (hunk 2 above)
Comment (Akreddy123) [reply src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1]: We are checking

File: src/AzSK/Framework/Abstracts/AzSKRoot.ps1 (hunk 4 above)
[the remark this answers is not in the document]
Comment (Akreddy123) [reply src/AzSK/Framework/Abstracts/AzSKRoot.ps1]: done

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 (hunk 3 above)
[the remark this answers is not in the document]
Comment (Akreddy123) [reply src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: Removed

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 (hunk 5 above)
Comment (Akreddy123) [reply src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: Removed the hardcoding

File: src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1 [hunk 6]
```diff
@@ -2987,6 +2987,19 @@ class CCAutomation: CommandBase
...
+		{
+			$AzureEnv = "AzureCloud"
+		}
+		try
+		{
+		$ManagementUri = [Helpers]::GetCurrentRMContext().Environment.ServiceManagementUrl
```
Comment (Akreddy123) [inline src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1]: move to common helper fn and do the logic only if env  is not azure cloud

File: src/AzSK/Framework/Helpers/Helpers.ps1 [hunk 7]
```diff
@@ -15,12 +20,25 @@ class Helpers {
...
 			if ((-not $rmContext) -or ($rmContext -and (-not $rmContext.Subscription -or -not $rmContext.Account))) {
 				[EventBase]::PublishGenericCustomMessage("No active Azure login session found. Initiating login flow...", [MessageType]::Warning);
-
-				$rmLogin = Connect-AzureRmAccount
+                [PSObject]$rmLogin = $null
+                $AzureEnvironment= [Helpers]::AzureSettings.AzureEnvironment
+                if(-not [string]::IsNullOrWhiteSpace($AzureEnvironment))
+                {
+                    try{
+                        $rmLogin = Connect-AzureRmAccount -EnvironmentName AzureUSGovernment
```
Comment (Akreddy123) [inline src/AzSK/Framework/Helpers/Helpers.ps1]: correct variable name

File: src/AzSK/Framework/Helpers/Helpers.ps1 [hunk 8]
```diff
@@ -5,8 +5,13 @@ using namespace Microsoft.Azure.Management.Storage.Models
 Set-StrictMode -Version Latest
 class Helpers {
 
-	static hidden [PSObject] $currentRMContext;
-
+    static hidden [PSObject] $currentRMContext;
+    static hidden [PSObject] $AzureSettings;
```
Comment (Akreddy123) [inline src/AzSK/Framework/Helpers/Helpers.ps1]: remove the variable

File: src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1 (hunk 6 above)
Comment (VishalHaibatpure) [reply src/AzSK/Framework/Core/ContinuousAssurance/CAAutomation.ps1]: Use [WebRequestHelper]::GetServiceManagementUrl()

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 [hunk 9]
```diff
@@ -465,11 +475,31 @@ class AppService: SVTBase
...
+					{
+						$temp = @($this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('azurewebsites') })
+						$AppURL = $this.FormatURL($temp[0])
+						$apiFunctionsUrl = [string]::Format("https://{0}.scm.{1}/api/functions",$this.ResourceContext.ResourceName,$AppURL)
+					}
+			    }
+				else
+				{
+					$temp = @($this.ResourceObject.Properties.HostNames | where-object { $_.Contains('azurewebsites') })
+					$AppURL = $this.FormatURL($temp[0])
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: Can we avoid duplicate code?

File: src/AzSK/Framework/Core/SVT/Services/AppService.ps1 [hunk 10]
```diff
@@ -586,12 +616,32 @@ class AppService: SVTBase
...
+		{
+			if((($this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('scm') }) | Measure-Object).Count -eq 1)
+			{
+				$scmURL = $this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('scm') }
+				$apiFunctionsUrl = [string]::Format("https://{0}/api/functions",$scmURL)
+			}
+			else
+			{
+				$temp = @($this.WebAppDetails.EnabledHostNames | where-object { $_.Contains('azurewebsites') })
+				$AppURL = $this.FormatURL($temp[0])
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Core/SVT/Services/AppService.ps1]: Avoid duplicate code in multiple functions

File: src/AzSK/Framework/Helpers/ConfigurationHelper.ps1 [hunk 11]
```diff
@@ -268,9 +268,17 @@ class ConfigurationHelper {
 		[System.Uri] $validatedUri = $null;
 		if([System.Uri]::TryCreate($uri, [System.UriKind]::Absolute, [ref] $validatedUri))
 		{
+			$rmContext = [Helpers]::GetCurrentRMContext();
+			if(-not [string]::IsNullOrWhiteSpace($rmContext.Environment.Name))
+		     {
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Helpers/ConfigurationHelper.ps1]: Is it possible to call [WebRequestHelper]::GetServiceManagementUrl() here ?

File: src/AzSK/Framework/Helpers/WebRequestHelper.ps1 [hunk 12]
```diff
@@ -13,7 +13,33 @@ class WebRequestHelper {
 	static [System.Object[]] InvokeGetWebRequest([string] $uri) 
 	{	
         return [WebRequestHelper]::InvokeGetWebRequest($uri, [WebRequestHelper]::GetAuthHeaderFromUri($uri));
-    }
+	}
+	
+	hidden static [string] GetResourceManagerUrl()
+	{
+		if( [string]::IsNullOrWhiteSpace([AzSKSettings]::GetInstance().AzureEnvironment))
```
Comment (VishalHaibatpure) [inline src/AzSK/Framework/Helpers/WebRequestHelper.ps1]: Just a optimization suggestion. We can combine both if statements. Something like below.

$azureEnv= [AzSKSettings]::GetInstance().AzureEnvironmen
if(-not  [string]::IsNullOrWhiteSpace($azureEnv) and -($azureEnv -ne "AzureCloud"))
{
 return [Helpers]::GetCurrentRMContext().Environment.ResourceManagerUrl 
}

**Pull request merged**